### PR TITLE
Support edge certificate rotation

### LINKED ISSUE #1630: Add certificate rotation for edge node
**Why is this needed**:
At present, the edge nodes of KubeEdge only apply for certificates when they first started.
Because the certificates have a validity period, we need to automatically generate a new key and request a new certificate from the cloudcore periodically.

### PR DESCRIPTION
**Pull request by XJangel**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Ref: #1630 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cloud/pkg/cloudhub/servers/httpserver/certutil.go [hunk 1]
```diff
@@ -114,7 +116,7 @@ func NewCertFromCa(cfg *certutil.Config, caCert *x509.Certificate, serverKey cry
 		IPAddresses:  cfg.AltNames.IPs,
 		SerialNumber: serial,
 		NotBefore:    time.Now().UTC(),
-		NotAfter:     time.Now().Add(time.Hour * 24 * 365 * 100),
+		NotAfter:     time.Now().Add(time.Hour * 24 * time.Duration(t)),
```
Comment (daixiang0) [inline cloud/pkg/cloudhub/servers/httpserver/certutil.go:119]: Here do we need to count how many days in a year, or just hardcode 365 days?

Comment (daixiang0) [reply cloud/pkg/cloudhub/servers/httpserver/certutil.go:119]: Oh, i see this config option `EdgeCertValidityPeriod` while not see where to call it, is it normal?

Comment (XJangel) [reply cloud/pkg/cloudhub/servers/httpserver/certutil.go:119]: > Here do we need to count how many days in a year, or just hardcode 365 days?

We don't count the days in a year. The certificate of CloudCore remains valid for 36500 days. While the certificate of EdgeCore can be manually set in `EdgeCertValidityPeriod` .

Comment (XJangel) [reply cloud/pkg/cloudhub/servers/httpserver/certutil.go:119]: > Oh, i see this config option `EdgeCertValidityPeriod` while not see where to call it, is it normal?

image
Here we call it to set the valid period of EdgeCore certificate.

Comment (daixiang0) [reply cloud/pkg/cloudhub/servers/httpserver/certutil.go:119]: Oh i am confused with `validalityPeriod` and `EdgeCertValidityPeriod`.

**Comment by XJangel** [comment]
> Thanks for the effort, leave some comments, ptal:) @XJangel

I get it and will update :)

File: edge/pkg/edgehub/certificate/certmanager.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+// nextRotationDeadline returns the rotation deadline. It is different in every rotation.
+func (cm *CertManager) nextRotationDeadline() (time.Time, error) {
+	cert, err := cm.getCurrent()
+	if err != nil {
+		return time.Time{}, fmt.Errorf("faild to get current certificate")
+	}
+	notAfter := cert.Leaf.NotAfter
+	totalDuration := float64(notAfter.Sub(cert.Leaf.NotBefore))
+	deadline := cert.Leaf.NotBefore.Add(jitteryDuration(totalDuration))
+	klog.Infof("Certificate expiration is %v, rotation deadline is %v", notAfter, deadline)
```
Comment (fisherxu) [inline edge/pkg/edgehub/certificate/certmanager.go:195]: Need add the log level

File: cloud/pkg/cloudhub/servers/httpserver/server.go [hunk 3]
```diff
@@ -93,23 +94,59 @@ func EncodeCertPEM(cert *x509.Certificate) []byte {
...
+		} else {
+			signEdgeCert(w, r)
+		}
+		return
+	}
+	if ok := verifyAuthorization(w, r); ok {
+		signEdgeCert(w, r)
+	} else {
+		klog.Errorf("failed to sign the certificate for edgenode: %s, invalid token", r.Header.Get("NodeName"))
+	}
```
Comment (fisherxu) [inline cloud/pkg/cloudhub/servers/httpserver/server.go:113]:
```suggestion
	if  !verifyAuthorization(w, r) {
		klog.Errorf("failed to sign the certificate for edgenode: %s, invalid token", r.Header.Get("NodeName"))
	} 
	signEdgeCert(w, r)
```

File: cloud/pkg/cloudhub/servers/httpserver/server.go [hunk 4]
```diff
@@ -118,30 +155,34 @@ func edgeCoreClientCert(w http.ResponseWriter, r *http.Request) {
...
 	}
+	return true
+}
 
+// signEdgeCert signs the CSR from EdgeCore
+func signEdgeCert(w http.ResponseWriter, r *http.Request) {
 	csrContent, err := ioutil.ReadAll(r.Body)
 	if err != nil {
-		klog.Errorf("fail to read file! error:%v", err)
+		klog.Errorf("fail to read file when signing the cert for edgenode:%s! error:%v", r.Header.Get("NodeName"), err)
```
Comment (fisherxu) [inline cloud/pkg/cloudhub/servers/httpserver/server.go:176]: For the "nodename", could we use a const?

File: edge/pkg/edgehub/certificate/certmanager.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+// rotate starts edge certificate rotation process
+func (cm *CertManager) rotate() {
+	klog.Infof("Certificate rotation is enabled.")
+	go wait.Forever(func() {
+		deadline, err := cm.nextRotationDeadline()
+		if err != nil {
+			klog.Errorf("failed to get next rotation deadline:%v", err)
+		}
+		if sleepInterval := deadline.Sub(cm.now()); sleepInterval > 0 {
+			klog.Infof("Waiting %v for next certificate rotation", sleepInterval)
```
Comment (fisherxu) [inline edge/pkg/edgehub/certificate/certmanager.go:165]: Add log level

File: cloud/pkg/cloudhub/servers/httpserver/server.go (hunk 3 above)
Comment (XJangel) [reply cloud/pkg/cloudhub/servers/httpserver/server.go:113]: If we do this, whether the token is right or not cloudcore will always sign the certificate for edgenode.

File: cloud/pkg/cloudhub/servers/httpserver/server.go (hunk 4 above)
Comment (XJangel) [reply cloud/pkg/cloudhub/servers/httpserver/server.go:176]: yep. Got it

File: cloud/pkg/cloudhub/servers/httpserver/server.go (hunk 3 above)
Comment (fisherxu) [reply cloud/pkg/cloudhub/servers/httpserver/server.go:113]: ohh, if we change as my comments that we need return in `if`. So keep your original code is fine :)

Comment (fisherxu) [reply cloud/pkg/cloudhub/servers/httpserver/server.go:113]: And seems that `ok` can be deleted.

Comment (XJangel) [reply cloud/pkg/cloudhub/servers/httpserver/server.go:113]: got it.

**Pull request merged**